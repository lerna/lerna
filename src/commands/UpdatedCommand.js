"use strict";

const _ = require("lodash");
const chalk = require("chalk");

const Command = require("../Command");
const output = require("../utils/output");
const publishOptions = require("./PublishCommand").builder;
const UpdatedPackagesCollector = require("../UpdatedPackagesCollector");

const updatedOptions = _.assign({}, publishOptions, {
  json: {
    describe: "Show information in JSON format",
    group: "Command Options:",
    type: "boolean",
    default: undefined,
  },
});

exports.handler = function handler(argv) {
  // eslint-disable-next-line no-use-before-define
  return new UpdatedCommand(argv);
};

exports.command = "updated";

exports.describe = "Check which packages have changed since the last publish.";

exports.builder = yargs => yargs.options(updatedOptions);

class UpdatedCommand extends Command {
  initialize(callback) {
    this.updates = new UpdatedPackagesCollector(this).getUpdates();

    const proceedWithUpdates = this.updates.length > 0;
    if (!proceedWithUpdates) {
      this.logger.info("No packages need updating");
    }

    callback(null, proceedWithUpdates, 1);
  }

  get otherCommandConfigs() {
    return ["publish"];
  }

  get defaultOptions() {
    return Object.assign({}, super.defaultOptions, {
      json: false,
    });
  }

  execute(callback) {
    const updatedPackages = this.updates.map(update => update.package).map(pkg => ({
      name: pkg.name,
      version: pkg.version,
      private: pkg.private,
    }));

    this.logger.info("result");
    if (this.options.json) {
      output(JSON.stringify(updatedPackages, null, 2));
    } else {
      const formattedUpdates = updatedPackages
        .map(pkg => `- ${pkg.name}${pkg.private ? ` (${chalk.red("private")})` : ""}`)
        .join("\n");
      output(formattedUpdates);
    }

    callback(null, true);
  }
}
