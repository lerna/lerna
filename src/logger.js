import progressBar from "./progressBar";
import chalk from "chalk";
import pad from "pad";

const cwd = process.cwd();

const DEFAULT_LOGLEVEL = "info";

const LEVELS = [
  [ "silly",   "magenta" ],
  [ "verbose", "blue"    ],
  [ "info",    "reset"   ],
  [ "success", "green"   ],
  [ "warn",    "yellow"  ],
  [ "error",   "red"     ],
  [ "silent",            ],
];

const TYPE_TO_LEVEL = LEVELS
  .reduce((map, [type], index) => (map[type] = index, map), {});

class Logger {
  constructor() {
    this.setLogLevel();
    this.logs = [];
  }

  setLogLevel(type) {
    this.loglevel = TYPE_TO_LEVEL[type || DEFAULT_LOGLEVEL];
  }

  _log(type, style, level, message, error) {
    this.logs.push({
      type,
      message,
      error
    });

    if (level < this.loglevel) {
      return;
    }

    if (error) {
      message += "\n" + (error.stack || error);
    }

    if (style) {
      message = style(message);
    }

    progressBar.clear();
    this._emit(message);
    progressBar.restore();
  }

  _emit(message) {
    if (process.env.NODE_ENV !== "lerna-test") {
      console.log(message);
    }
  }

  newLine() {
    this._emit("");
  }

  logifyAsync() {
    return (target, property, descriptor) => {
      const message = target.name + "." + property;
      const method = descriptor.value;

      descriptor.value = (...args) => {
        const callback = args.pop();
        const msg = this._formatMethod(message, args);

        this.verbose(msg);

        // wrap final callback
        args.push((error, value) => {
          if (error) {
            this.error(msg, error);
          } else {
            this.verbose(msg + " => " + this._formatValue(value));
          }

          callback(error, value);
        });

        method(...args);
      };
    };
  }

  logifySync() {
    return (target, property, descriptor) => {
      const message = target.name + "." + property;
      const method = descriptor.value;

      descriptor.value = (...args) => {
        const msg = this._formatMethod(message, args);

        this.verbose(msg);

        try {
          const result = method(...args);
          this.verbose(msg + " => " + this._formatValue(result));
          return result;
        } catch (error) {
          this.error(msg, error);
          throw error;
        }
      };
    };
  }

  _formatMethod(method, args) {
    return pad(method, 30, " ") + "(" + this._formatArguments(args) + ")";
  }

  _formatArguments(args) {
    const fullArgs = args.map(this._formatValue).join(", ");
    if (fullArgs.length > 100) {
      return fullArgs.slice(0, 100) + "...";
    } else {
      return fullArgs;
    }
  }

  _formatValue(arg) {
    if (typeof arg === "function") {
      return "function " + arg.name + "() {...}";
    }

    return (JSON.stringify(arg) || "").replace(cwd, ".");
  }
}

LEVELS.forEach(([type, color]) => {
  if (!color) return; // "silent"
  const style = chalk[color];
  const level = TYPE_TO_LEVEL[type];
  Logger.prototype[type] = function(message, error) {
    this._log(type, style, level, message, error);
  };
});

export default new Logger();
