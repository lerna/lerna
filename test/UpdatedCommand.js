import chalk from "chalk";
import execa from "execa";
import path from "path";
import normalizeNewline from "normalize-newline";

// mocked or stubbed modules
import logger from "../src/logger";

// helpers
import exitWithCode from "./helpers/exitWithCode";
import initFixture from "./helpers/initFixture";
import updateLernaConfig from "./helpers/updateLernaConfig";

// file under test
import UpdatedCommand from "../src/commands/UpdatedCommand";

// keep snapshots stable cross-platform
chalk.enabled = false;

const normalized = (spy) =>
  spy.mock.calls.map((args) => normalizeNewline(args[0]));

const gitTag = (opts) => execa.sync("git", ["tag", "v1.0.0"], opts);
const gitAdd = (opts) => execa.sync("git", ["add", "-A"], opts);
const gitCommit = (opts) => execa.sync("git", ["commit", "-m", "Commit"], opts);
const touchFile = (opts) => (filePath) =>
    execa.sync("touch", [path.join(opts.cwd, filePath)], opts);

const setupGitChanges = (testDir, filePaths) => {
  const opts = {
    cwd: testDir,
  };
  gitTag(opts);
  filePaths.forEach(touchFile(opts));
  gitAdd(opts);
  gitCommit(opts);
};

// isolates singleton logger method from other command instances
const stubLogger = (instance, logMethod) =>
  jest.spyOn(instance.logger, logMethod).mockImplementation(() => {});

describe("UpdatedCommand", () => {
  const loggerInfo = logger.info;

  afterEach(() => {
    logger.info = loggerInfo;
  });

  /** =========================================================================
   * Basic
   * ======================================================================= */

  describe("basic", () => {
    let testDir;

    beforeEach(() => initFixture("UpdatedCommand/basic").then((dir) => {
      testDir = dir;
    }));

    it("should list changes", (done) => {
      setupGitChanges(testDir, [
        "packages/package-2/random-file",
      ]);

      const updatedCommand = new UpdatedCommand([], {});

      updatedCommand.runValidations();
      updatedCommand.runPreparations();

      const logInfo = stubLogger(updatedCommand, "info");

      updatedCommand.runCommand(exitWithCode(0, (err) => {
        if (err) return done.fail(err);
        try {
          expect(normalized(logInfo)).toMatchSnapshot();
          done();
        } catch (ex) {
          done.fail(ex);
        }
      }));
    });

    it("should list all packages when no tag is found", (done) => {
      const updatedCommand = new UpdatedCommand([], {});

      updatedCommand.runValidations();
      updatedCommand.runPreparations();

      const logInfo = stubLogger(updatedCommand, "info");

      updatedCommand.runCommand(exitWithCode(0, (err) => {
        if (err) return done.fail(err);
        try {
          expect(normalized(logInfo)).toMatchSnapshot();
          done();
        } catch (ex) {
          done.fail(ex);
        }
      }));
    });

    it("should list changes with --force-publish *", (done) => {
      setupGitChanges(testDir, [
        "packages/package-2/random-file",
      ]);

      const updatedCommand = new UpdatedCommand([], {
        forcePublish: "*"
      });

      updatedCommand.runValidations();
      updatedCommand.runPreparations();

      const logInfo = stubLogger(updatedCommand, "info");

      updatedCommand.runCommand(exitWithCode(0, (err) => {
        if (err) return done.fail(err);
        try {
          expect(normalized(logInfo)).toMatchSnapshot();
          done();
        } catch (ex) {
          done.fail(ex);
        }
      }));
    });

    it("should list changes with --force-publish [package,package]", (done) => {
      setupGitChanges(testDir, [
        "packages/package-3/random-file",
      ]);

      const updatedCommand = new UpdatedCommand([], {
        forcePublish: "package-2,package-4"
      });

      updatedCommand.runValidations();
      updatedCommand.runPreparations();

      const logInfo = stubLogger(updatedCommand, "info");

      updatedCommand.runCommand(exitWithCode(0, (err) => {
        if (err) return done.fail(err);
        try {
          expect(normalized(logInfo)).toMatchSnapshot();
          done();
        } catch (ex) {
          done.fail(ex);
        }
      }));
    });

    it("should list changes without ignored files", (done) => {
      updateLernaConfig(testDir, {
        commands: { // "command" also supported
          publish: {
            ignore: ["ignored-file"],
          },
        },
      });

      setupGitChanges(testDir, [
        "packages/package-2/ignored-file",
        "packages/package-3/random-file",
      ]);

      const updatedCommand = new UpdatedCommand([], {});

      updatedCommand.runValidations();
      updatedCommand.runPreparations();

      const logInfo = stubLogger(updatedCommand, "info");

      updatedCommand.runCommand(exitWithCode(0, (err) => {
        if (err) return done.fail(err);
        try {
          expect(normalized(logInfo)).toMatchSnapshot();
          done();
        } catch (ex) {
          done.fail(ex);
        }
      }));
    });

    it("should list changes for explicitly changed packages", (done) => {
      setupGitChanges(testDir, [
        "packages/package-2/random-file",
      ]);

      const updatedCommand = new UpdatedCommand([], {
        [SECRET_FLAG]: true
      });

      updatedCommand.runValidations();
      updatedCommand.runPreparations();

      const logInfo = stubLogger(updatedCommand, "info");

      updatedCommand.runCommand(exitWithCode(0, (err) => {
        if (err) return done.fail(err);
        try {
          expect(normalized(logInfo)).toMatchSnapshot();
          done();
        } catch (ex) {
          done.fail(ex);
        }
      }));
    });

    it("should list changes in private packages", (done) => {
      setupGitChanges(testDir, [
        "packages/package-5/random-file",
      ]);

      const updatedCommand = new UpdatedCommand([], {});

      updatedCommand.runValidations();
      updatedCommand.runPreparations();

      const logInfo = stubLogger(updatedCommand, "info");

      updatedCommand.runCommand(exitWithCode(0, (err) => {
        if (err) return done.fail(err);
        try {
          expect(normalized(logInfo)).toMatchSnapshot();
          done();
        } catch (ex) {
          done.fail(ex);
        }
      }));
    });
  });

  /** =========================================================================
   * Circular
   * ======================================================================= */

  describe("circular", () => {
    let testDir;

    beforeEach(() => initFixture("UpdatedCommand/circular").then((dir) => {
      testDir = dir;
    }));

    it("should list changes", (done) => {
      setupGitChanges(testDir, [
        "packages/package-3/random-file",
      ]);

      const updatedCommand = new UpdatedCommand([], {});

      updatedCommand.runValidations();
      updatedCommand.runPreparations();

      const logInfo = stubLogger(updatedCommand, "info");

      updatedCommand.runCommand(exitWithCode(0, (err) => {
        if (err) return done.fail(err);
        try {
          expect(normalized(logInfo)).toMatchSnapshot();
          done();
        } catch (ex) {
          done.fail(ex);
        }
      }));
    });

    it("should list changes with --force-publish *", (done) => {
      setupGitChanges(testDir, [
        "packages/package-2/random-file",
      ]);

      const updatedCommand = new UpdatedCommand([], {
        forcePublish: "*"
      });

      updatedCommand.runValidations();
      updatedCommand.runPreparations();

      const logInfo = stubLogger(updatedCommand, "info");

      updatedCommand.runCommand(exitWithCode(0, (err) => {
        if (err) return done.fail(err);
        try {
          expect(normalized(logInfo)).toMatchSnapshot();
          done();
        } catch (ex) {
          done.fail(ex);
        }
      }));
    });

    it("should list changes with --force-publish [package,package]", (done) => {
      setupGitChanges(testDir, [
        "packages/package-4/random-file",
      ]);

      const updatedCommand = new UpdatedCommand([], {
        forcePublish: "package-2"
      });

      updatedCommand.runValidations();
      updatedCommand.runPreparations();

      const logInfo = stubLogger(updatedCommand, "info");

      updatedCommand.runCommand(exitWithCode(0, (err) => {
        if (err) return done.fail(err);
        try {
          expect(normalized(logInfo)).toMatchSnapshot();
          done();
        } catch (ex) {
          done.fail(ex);
        }
      }));
    });

    it("should list changes without ignored files", (done) => {
      updateLernaConfig(testDir, {
        command: { // "commands" also supported
          publish: {
            ignore: ["ignored-file"],
          },
        },
      });

      setupGitChanges(testDir, [
        "packages/package-2/ignored-file",
        "packages/package-3/random-file",
      ]);

      const updatedCommand = new UpdatedCommand([], {});

      updatedCommand.runValidations();
      updatedCommand.runPreparations();

      const logInfo = stubLogger(updatedCommand, "info");

      updatedCommand.runCommand(exitWithCode(0, (err) => {
        if (err) return done.fail(err);
        try {
          expect(normalized(logInfo)).toMatchSnapshot();
          done();
        } catch (ex) {
          done.fail(ex);
        }
      }));
    });
  });
});

const SECRET_FLAG = new Buffer("ZGFuZ2Vyb3VzbHlPbmx5UHVibGlzaEV4cGxpY2l0VXBkYXRlc1RoaXNJc0FDdXN0b21GbGFnRm9yQmFiZWxBbmRZb3VTaG91bGROb3RCZVVzaW5nSXRKdXN0RGVhbFdpdGhNb3JlUGFja2FnZXNCZWluZ1B1Ymxpc2hlZEl0SXNOb3RBQmlnRGVhbA==", "base64").toString("ascii");
