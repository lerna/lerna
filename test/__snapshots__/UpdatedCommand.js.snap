// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdatedCommand basic should list all packages when no tag is found 1`] = `
Array [
  "Checking for updated packages...",
  "No tags found! Comparing with initial commit.",
  "- package-1
- package-2
- package-3
- package-4
- package-5 (private)",
]
`;

exports[`UpdatedCommand basic should list changes 1`] = `
Array [
  "Checking for updated packages...",
  "Comparing with: v1.0.0",
  "- package-2
- package-3",
]
`;

exports[`UpdatedCommand basic should list changes for explicitly changed packages 1`] = `
Array [
  "Checking for updated packages...",
  "Comparing with: v1.0.0",
  "- package-2",
]
`;

exports[`UpdatedCommand basic should list changes in private packages 1`] = `
Array [
  "Checking for updated packages...",
  "Comparing with: v1.0.0",
  "- package-5 (private)",
]
`;

exports[`UpdatedCommand basic should list changes with --force-publish * 1`] = `
Array [
  "Checking for updated packages...",
  "Comparing with: v1.0.0",
  "- package-1
- package-2
- package-3
- package-4
- package-5 (private)",
]
`;

exports[`UpdatedCommand basic should list changes with --force-publish [package,package] 1`] = `
Array [
  "Checking for updated packages...",
  "Comparing with: v1.0.0",
  "- package-2
- package-3
- package-4",
]
`;

exports[`UpdatedCommand basic should list changes without ignored files 1`] = `
Array [
  "Checking for updated packages...",
  "Comparing with: v1.0.0",
  "- package-3",
]
`;

exports[`UpdatedCommand circular should list changes 1`] = `
Array [
  "Checking for updated packages...",
  "Comparing with: v1.0.0",
  "- package-3
- package-4",
]
`;

exports[`UpdatedCommand circular should list changes with --force-publish * 1`] = `
Array [
  "Checking for updated packages...",
  "Comparing with: v1.0.0",
  "- package-1
- package-2
- package-3
- package-4
- package-5 (private)",
]
`;

exports[`UpdatedCommand circular should list changes with --force-publish [package,package] 1`] = `
Array [
  "Checking for updated packages...",
  "Comparing with: v1.0.0",
  "- package-2
- package-3
- package-4",
]
`;

exports[`UpdatedCommand circular should list changes without ignored files 1`] = `
Array [
  "Checking for updated packages...",
  "Comparing with: v1.0.0",
  "- package-3
- package-4",
]
`;
